### PR TITLE
Solution for the print issue

### PR DESCRIPTION
**Pull request by mansogf**
To run the project, please, use the python 3.

Is expected that the program could print a tree with 17 random nodes. This number is in consequence of my terminal size, but if you have a big screen you can test it with more nodes.

I hope you like it and again, thank you very much for this project! It has been a pleasure to have the opportunity to contribute with this.

### CODE REVIEW

File: rb_tree.py [hunk 1]
```diff
@@ -523,3 +534,100 @@ def _get_sibling(self, node):
...
+        if root is None:
+            return [], 0, 0, 0
+
+        line1 = []
+        line2 = []
+        
+        if index:
+            node_repr = '{}{}{}'.format(curr_index, delimiter, root.value)
+        else:
+            node_repr = str(root.value)
```
Comment (stanislavkozlovski) [inline rb_tree.py]: nit: Mind just doing `'NIL' if root.value is None else str(root.value)`? Results in more succinct code

File: rb_tree.py [hunk 2]
```diff
@@ -523,3 +534,100 @@ def _get_sibling(self, node):
...
+
+"""
+
+Testing the print with a tree with 17 random nodes
+
+"""
+
+import numpy as np
+
+arvore = RedBlackTree()
```
Comment (stanislavkozlovski) [inline rb_tree.py]: What language is this variable name in? We should use english

File: rb_tree.py [hunk 3]
```diff
@@ -523,3 +534,100 @@ def _get_sibling(self, node):
...
+        len_new_box = len(new_box[0])
+        if root.color == 'RED':
+            len_new_box = len(new_box[0]) - 16
+
+        # Return the new box, its width and its root repr positions
+        return new_box, len_new_box, new_root_start, new_root_end
+
+"""
+
+Testing the print with a tree with 17 random nodes
```
Comment (stanislavkozlovski) [inline rb_tree.py]: I think we should move this into a unit test to not mix it up with the main code

File: rb_tree.py [hunk 4]
```diff
@@ -523,3 +534,100 @@ def _get_sibling(self, node):
             sibling = parent.right
             direction = 'R'
         return sibling, direction
+
+    def _build_tree_string(self, root=None, curr_index=0, index=False, delimiter='-'):
+
```
Comment (stanislavkozlovski) [inline rb_tree.py]: Could we also add the docstring from the original source?

File: rb_tree.py [hunk 5]
```diff
@@ -9,48 +9,55 @@
...
         self.left = left
         self.right = right
 
     def __repr__(self):
+        
         return '{color} {val} Node'.format(color=self.color, val=self.value)
 
     def __iter__(self):
+        
+        print(str(self.value) + " (" + str(self.color)+")")
```
Comment (stanislavkozlovski) [inline rb_tree.py]: I don't think we should have the print statement be in the `__iter__()` method. You do not always want to print

File: rb_tree.py [hunk 6]
```diff
@@ -9,48 +9,55 @@
...
         if self.right.color != NIL:
+            print('right ({})'.format(str(self.value)));
             yield from self.right.__iter__()
+ 
 
     def __eq__(self, other):
-        if self.color == NIL and self.color == other.color:
-            return True
+            
+            if self.color == NIL and self.color == other.color:
```
Comment (stanislavkozlovski) [inline rb_tree.py]: Does this even compile? There is too much space (8 spaces) in this method body

**Review by stanislavkozlovski — COMMENTED** [review]
Thanks for the PR @gabrielfilipe7unb!
I love the initiative and this change will be very helpful to visualize the tree.

I have a couple of blockers for merging this:
1. The formatting is weird. We have one method with 8 spaces and I don't think that even compiles
2. It is missing PEP8 validation. There some changes (like adding a new line to the beginning of a method) which does not pass the Python standard PEP8 ruleset. Please make sure the code passes the PEP8 standard - this is used throughout most Python codebases
http://pep8online.com/

File: rb_tree.py (hunk 2 above)
Comment (mansogf) [reply rb_tree.py]: Fixed.

File: rb_tree.py (hunk 6 above)
Comment (mansogf) [reply rb_tree.py]: It was compiling but it was a mistake. Now, fixed.

File: rb_tree.py (hunk 5 above)
Comment (mansogf) [reply rb_tree.py]: Sure! It was part of some tests that I was doing. Now, it was deleted. Thanks!

File: rb_tree.py (hunk 4 above)
Comment (mansogf) [reply rb_tree.py]: Ok, done!

File: rb_tree.py (hunk 3 above)
Comment (mansogf) [reply rb_tree.py]: Ok! Done.

File: rb_tree.py (hunk 1 above)
Comment (mansogf) [reply rb_tree.py]: Done! Check it out please.

**Comment by mansogf** [comment]
I believe that all the requests were fixed now. Please, just check it out one more time! 

Thank you very much!

File: rb_tree.py [hunk 7]
```diff
@@ -35,29 +37,34 @@ def __eq__(self, other):
...
-        """ Returns the number of NOT NIL children the node has """
         if self.color == NIL:
             return 0
-        return sum([int(self.left.color != NIL), int(self.right.color != NIL)])
+        return (sum([int(self.left.color != NIL),
+                int(self.right.color != NIL)]))
 
 
 class RedBlackTree:
-    # every node has null nodes as children initially, create one such object for easy management
```
Comment (stanislavkozlovski) [inline rb_tree.py:53]: Why remove this comment?

File: rb_tree.py [hunk 8]
```diff
@@ -35,29 +37,34 @@ def __eq__(self, other):
...
+            other.parent.value and self.parent.color == other.parent.color
+        return (self.value == other.value and
+                self.color == other.color and parents_are_same)
 
     def has_children(self) -> bool:
-        """ Returns a boolean indicating if the node has children """
         return bool(self.get_children_count())
 
     def get_children_count(self) -> int:
-        """ Returns the number of NOT NIL children the node has """
```
Comment (stanislavkozlovski) [inline rb_tree.py:46]: Why remove this comment?

File: rb_tree.py [hunk 9]
```diff
@@ -35,29 +37,34 @@ def __eq__(self, other):
...
-            parents_are_same = self.parent.value == other.parent.value and self.parent.color == other.parent.color
-        return self.value == other.value and self.color == other.color and parents_are_same
+            parents_are_same = self.parent.value
+            self.parent.value == other.parent.value and self.parent.color
+            other.parent.value and self.parent.color == other.parent.color
+        return (self.value == other.value and
+                self.color == other.color and parents_are_same)
 
     def has_children(self) -> bool:
-        """ Returns a boolean indicating if the node has children """
```
Comment (stanislavkozlovski) [inline rb_tree.py]: Why remove this comment?

File: rb_tree.py [hunk 10]
```diff
@@ -35,29 +37,34 @@ def __eq__(self, other):
         if self.parent is None or other.parent is None:
             parents_are_same = self.parent is None and other.parent is None
         else:
-            parents_are_same = self.parent.value == other.parent.value and self.parent.color == other.parent.color
```
Comment (stanislavkozlovski) [inline rb_tree.py:38]: Let's not alter the formatting of unrelated code for this PR specifically

File: rb_tree.py [hunk 11]
```diff
@@ -21,11 +23,11 @@ def __repr__(self):
 
     def __iter__(self):
         if self.left.color != NIL:
+            print('left ({})'.format(str(self.value)))
             yield from self.left.__iter__()
 
-        yield self.value
```
Comment (stanislavkozlovski) [inline rb_tree.py:26]: This will be a bug - we should not remove this line at all.

**Review by stanislavkozlovski — CHANGES_REQUESTED** [review]
Thanks again for the work on this.

Could you probably start the PR from scratch again or go over every change you've made? There are quite a lot of changes in formatting of the code which shouldn't be there - we're looking at adding a new print method and a test only. It's good practice to keep the scope of pull requests as small as possible such that we don't risk introducing new bugs into the code.

File: rb_tree.py (hunk 7 above)
Comment (mansogf) [reply rb_tree.py:53]: Actually I don't realized that it was removed. I'll put all these comments in their places again.

File: rb_tree.py (hunk 9 above)
Comment (mansogf) [reply rb_tree.py]: Actually I don't realized that it was removed. I'll put all these comments in their places again.

File: rb_tree.py (hunk 11 above)
Comment (mansogf) [reply rb_tree.py:26]: Ok!

File: rb_tree.py (hunk 10 above)
Comment (mansogf) [reply rb_tree.py:38]: I just changed this specific line to fix the pep8 problem. But now, It was changed and I believe that is working perfectly.

File: rb_tree.py [hunk 12]
```diff
@@ -9,6 +10,7 @@
 
 
 class Node:
+
```
Comment (stanislavkozlovski) [inline rb_tree.py:13]: Again, this empty line is not needed

File: rb_tree.py [hunk 13]
```diff
@@ -21,12 +23,13 @@ def __repr__(self):
...
         if self.left.color != NIL:
+            print('left ({})'.format(str(self.value)))
             yield from self.left.__iter__()
 
-        yield self.value
-
         if self.right.color != NIL:
+            print('right ({})'.format(str(self.value)))
             yield from self.right.__iter__()
+        yield self.value
```
Comment (stanislavkozlovski) [inline rb_tree.py:32]: We shouldn't move this yield statement, it changes the logic

File: rb_tree.py (hunk 8 above)
Comment (stanislavkozlovski) [reply rb_tree.py:46]: This isn't resolved, the comment is still missing

File: rb_tree.py [hunk 14]
```diff
@@ -35,29 +38,39 @@ def __eq__(self, other):
         if self.parent is None or other.parent is None:
             parents_are_same = self.parent is None and other.parent is None
         else:
-            parents_are_same = self.parent.value == other.parent.value and self.parent.color == other.parent.color
-        return self.value == other.value and self.color == other.color and parents_are_same
+            parents_are_same = self.parent.value == (
+             other.parent.value and self.parent.color) == other.parent.color
+
+        return (self.value == other.value and
```
Comment (stanislavkozlovski) [inline rb_tree.py:44]: Could we please not change unnecessary stuff?

**Review by stanislavkozlovski — CHANGES_REQUESTED** [review]
I'm sorry Gabriel, but I again strongly suggest you start off from a new branch and implement just the necessary changes for resolving this issue. I didn't go through all the changes but I see that we still suffer from the same thing

If you open the "File Changed" tab (https://github.com/stanislavkozlovski/Red-Black-Tree/pull/5/files), you'll see that there are a ton of changes almost none of which are related to the print functionality.

This is the third time I've reviewed this and it hasn't been addressed. I want to get this merged but I don't have the time to repeat myself.

File: rb_tree.py (hunk 14 above)
Comment (mansogf) [reply rb_tree.py:44]: It was changed in consequence of the PEP8 request. It exceeds the number of characters per line.
